src/dmd/dcast.d

```diff
@@ -72,6 +72,23 @@ Expression implicitCastTo(Expression e, Scope* sc, Type t)
             MATCH match = e.implicitConvTo(t);
             if (match)
             {
+                // if the matching was done through alias this, do not cast
+                // the base type
+                auto ad = isAggregate(e.type);
+                if (ad && ad.aliasthis)
+                {
+                    auto aliasType = ad.type.aliasthisOf();
+                    MATCH aliasMatch = aliasType.implicitConvTo(t);
+                    // if alias has the same matching level as e
+                    // but their types are different, then alias this was matched
```

Comment:
This just seems awfully brittle (second-guessing what implicitConvTo() may have done). Secondly, it doesn't take into consideration structs. I think any fix for this should go into implicitConvTo().

Reply:
I fixed it now.